client/src/AppBundle/Service/DocumentSyncService.php

```diff
@@ -24,87 +29,223 @@ class DocumentSyncService
+        $s3Response = $this->retrieveDocumentContentFromS3($document);
 
-            $data = json_decode(strval($response->getBody()), true);
-            return $data['data']['id'];
-        } catch (RequestException $exception) {
-            $response = $exception->getResponse();
-            if ($response) {
-                $response = json_decode($exception->getResponse()->getBody());
+        if ($s3Response instanceof S3Exception) {
+            $syncStatus = in_array($s3Response->getAwsErrorCode(), S3Storage::MISSING_FILE_AWS_ERROR_CODES) ?
+                Document::SYNC_STATUS_PERMANENT_ERROR : Document::SYNC_STATUS_TEMPORARY_ERROR;
 
-                return $response->errors[0]->detail;
-            } else {
-                return $exception->getMessage();
-            }
+            $errorMessage = sprintf('S3 error: %s', $s3Response->getMessage());
+            return $this->handleDocumentStatusUpdate($document, $syncStatus, $errorMessage);
+        }
+
+        $siriusResponse = $this->handleSiriusSync($document, $s3Response);
+
+        if ($siriusResponse instanceof Throwable) {
+            $body = $siriusResponse->getResponse() ?
+                (string) $siriusResponse->getResponse()->getBody() : (string) $siriusResponse->getMessage();
+
+            // Add check for perm errors here and pass either temp or perm accordingly
+
+            return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_PERMANENT_ERROR, $body);
         }
+
+        $data = json_decode(strval($siriusResponse->getBody()), true);
+        $relevantSubmission = $report->getReportSubmissionByDocument($document);
+
+        $apiSubmissionResponse = $this->handleReportSubmissionUpdate($relevantSubmission->getId(), $data['data']['id']);
+
+        if ($apiSubmissionResponse instanceof Throwable) {
+            return null;
+        }
+
+        $apiDocumentResponse = $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_SUCCESS);
+
+        if ($apiDocumentResponse instanceof Throwable) {
+            return null;
+        }
+
+        return $apiDocumentResponse;
     }
 
     /**
-     * @param Report|Ndr $report
-     * @return array
+     * @param Document $document
+     * @return string|null
      */
-    private function getAttributesFromReport(ReportInterface $report): array
+    public function syncSupportingDocument(Document $document): ?string
     {
-        if ($report instanceof Ndr) {
-            $type = 'NDR';
-        } else if (in_array($report->getType(), [Report::TYPE_HEALTH_WELFARE, Report::TYPE_COMBINED_HIGH_ASSETS, Report::TYPE_COMBINED_LOW_ASSETS])) {
-            $type = 'HW';
+        if (!$document->supportingDocumentCanBeSynced()) {
+            return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_QUEUED);
+        }
+
+        $content = $this->retrieveDocumentContentFromS3($document);
+
+        if ($content instanceof Throwable) {
+            return null;
```

Comment:
This should use `handleDocumentStatusUpdate` to set status/error message like the report PDF does. Might be worth including that setting in `retrieveDocumentContentFromS3`, `handleSiriusSync` rather than handling it in the sync functions?

Alternatively, something like this (which I prefer because I'm not a fan of returning errors from functions):
```
try {
  $content = $this->retrieveDocumentContentFromS3($document);
  $this->handleSiriusSync($document, $content);
  return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_SUCCESS);
} catch (Throwable $e) {
  return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_PERMANENT_ERROR, $e->getMessage());
}
```

Reply:
I remember the day I was working on the error handling I was having an existential crisis about returning or throwing errors. I suspect I ended up on this solution as a temporary one and never came back to sorting it properly. I like your suggestion though so will go with that 👍

Reply:
Looks great 👍